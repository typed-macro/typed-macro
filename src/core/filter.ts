import { posix, win32 } from 'path'
import pm from 'picomatch'

export type FilterPattern = RegExp | string | Array<RegExp | string>

export type FilterOptions = {
  /**
   * Regexp or glob patterns to include matches, applied after
   * {@link FilterOptions.exclude}.
   *
   * The default value is `/\.[jt]sx?$/`.
   *
   * Once this option is defined, the default one will be overwritten.
   */
  include?: FilterPattern
  /**
   * Regexp or glob patterns to exclude matches, applied before
   * {@link FilterOptions.include}.
   *
   * The default value is `/node_modules/`.
   *
   * Once this option is defined, the default one will be overwritten.
   */
  exclude?: FilterPattern
}

export function createFilter({
  include = /\.[jt]sx?$/,
  exclude = /node_modules/,
}: FilterOptions) {
  const getMatcher = (pattern: string | RegExp) => {
    if (pattern instanceof RegExp) return pattern
    const fn = pm(pattern, { dot: true })
    return { test: (id: string) => fn(id) }
  }
  const ensureArray = (v: unknown) => (v ? (Array.isArray(v) ? v : [v]) : [])

  const includeMatchers = ensureArray(include).map(getMatcher)
  const hasIncludeRule = includeMatchers.length > 0
  const excludeMatchers = ensureArray(exclude).map(getMatcher)

  return (id: string) => {
    const pathId = id.split(win32.sep).join(posix.sep)
    for (let i = 0; i < excludeMatchers.length; ++i) {
      if (excludeMatchers[i].test(pathId)) return false
    }

    for (let i = 0; i < includeMatchers.length; ++i) {
      if (includeMatchers[i].test(pathId)) return true
    }

    return !hasIncludeRule
  }
}
