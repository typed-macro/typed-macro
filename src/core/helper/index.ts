export { BABEL_TOOLS, Babel } from './babel'
export { ImportOption } from './import'
export { getPathHelper, PathHelper } from './path'
export { getStateHelper, StateHelper } from './state'
export { getTransformHelper, TransformHelper } from './transform'
