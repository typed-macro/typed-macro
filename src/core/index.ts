export { Runtime, RuntimeOptions } from './runtime'
export {
  createTransformer,
  Transformer,
  TransformerOptions,
} from './transformer'
export {
  Macro,
  MacroContext,
  MacroHelper,
  MacroHandler,
  Exportable,
  NamespacedExportable,
  NormalizedExports,
  normalizeExports,
} from './types'
