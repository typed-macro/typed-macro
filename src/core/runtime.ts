import {
  createTypeRenderer,
  TypeRenderer,
  TypeRendererOptions,
} from '@/core/typeRenderer'
import {
  createTransformer,
  Transformer,
  TransformerOptions,
} from '@/core/transformer'
import {
  assertNoConflictMacro,
  NamespacedMacros,
  NamespacedModules,
  NamespacedTypes,
  NormalizedExports,
} from './exports'
import { findDuplicatedItem } from '@/common'

export type RuntimeOptions = {
  transformer: TransformerOptions
  typeRenderer: Pick<TypeRendererOptions, 'typesPath'>
}

export class Runtime {
  private macros: NamespacedMacros = Object.create(null)
  private modules: NamespacedModules = Object.create(null)
  private types: NamespacedTypes = Object.create(null)

  private macrosNamespaces: string[] = []
  private modulesNamespaces: string[] = []

  private devMode = false

  private readonly transformer: Transformer
  private readonly _typeRenderer: TypeRenderer

  constructor(options: RuntimeOptions) {
    this.transformer = createTransformer(options.transformer)
    this._typeRenderer = createTypeRenderer({
      types: this.types,
      typesPath: options.typeRenderer.typesPath,
    })
  }

  setDevMode(dev = true) {
    this.devMode = dev
  }

  register({ macros, modules, types }: NormalizedExports) {
    assertNoDuplicatedNamespace(Object.keys(this.types), Object.keys(types))
    assertNoDuplicatedNamespace(this.macrosNamespaces, Object.keys(macros))
    assertNoDuplicatedNamespace(this.modulesNamespaces, Object.keys(modules))
    assertNoConflictMacro(macros)
    Object.assign(this.macros, macros)
    this.macrosNamespaces = Object.keys(this.macros)
    Object.assign(this.modules, modules)
    this.modulesNamespaces = Object.keys(this.modules)
    Object.assign(this.types, types)
  }

  handleLoad(id: string) {
    if (this.macrosNamespaces.includes(id)) return 'export {}'
    if (this.modulesNamespaces.includes(id)) return this.modules[id]
  }

  handleResolveId(id: string) {
    if (
      this.macrosNamespaces.includes(id) ||
      this.modulesNamespaces.includes(id)
    )
      return id
  }

  handleTransform(code: string, filepath: string, ssr = false) {
    if (/\.[jt]sx?$/.test(filepath))
      return this.transformer(
        {
          code,
          filepath,
          ssr,
          dev: this.devMode,
        },
        this.macros
      )
  }

  get exports(): NormalizedExports {
    return {
      macros: this.macros,
      modules: this.modules,
      types: this.types,
    }
  }

  get typeRenderer() {
    return this._typeRenderer
  }
}

export function assertNoDuplicatedNamespace(ns1: string[], ns2: string[]) {
  const duplicated = findDuplicatedItem(ns1, ns2)
  if (duplicated) throw new Error(`duplicated namespace '${duplicated}'.`)
}
