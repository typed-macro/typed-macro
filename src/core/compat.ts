/*
 * This is a compatibility control mechanism based on version number,
 * to ensure the compatibility between macro and runtime.
 */

import { Macro } from '@/core/macro'
import { WithVersion } from '@/common'

/*
 * The current Macro version.
 *
 * It represents whether the call-context passed by the Runtime to the Macro is compatible.
 * This involves not only the structure of the call-context itself,
 * but also the structure of the AST between different versions of Babel,
 * for AST is generated in Runtime but transformed in Macros.
 *
 * @deps {@link RawMacroCall}, {@link Babel}
 */
export const MACRO_VERSION = 0

export function versionedMacro(v: Macro): Macro {
  ;(v as WithVersion<Macro>).$__version = MACRO_VERSION
  return v
}

export function isMacroCompatible(v: Macro) {
  return (v as WithVersion<Macro>).$__version === MACRO_VERSION
}
