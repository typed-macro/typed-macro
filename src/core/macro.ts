import { CallExpression } from '@babel/types'
import { NodePath } from '@babel/traverse'
import { Babel, PathHelper, StateHelper, TransformHelper } from '@/core/helper'

export type MacroContext = {
  /**
   * The file path of the module currently being handled.
   */
  filepath: string
  /**
   * The original source code of the module currently being handled.
   */
  code: string
  /**
   * The arguments nodes of the call-macro expression currently being handled.
   */
  args: CallExpression['arguments']
  /*
   * The NodePath of the call-macro expression currently being handled.
   */
  path: NodePath
  /*
   * Whether in ssr mode.
   */
  ssr: boolean
}

export type MacroHelper = TransformHelper & PathHelper & StateHelper

export type MacroHandler = (
  /**
   * Context about the macro caller.
   */
  ctx: MacroContext,
  /**
   * Babel tools to help operate node paths and nodes.
   */
  babel: Readonly<Babel>,
  /**
   * Wrappers upon babel tools to simplify operations.
   */
  helper: Readonly<MacroHelper>
) => void

export type Macro = {
  name: string
  apply: MacroHandler
}

export type MacroMeta = {
  signatures: {
    comment?: string
    signature: string
  }[]
  types: string[]
}

type InternalMacro = Macro & {
  __types: string
}

export function macro(
  name: string,
  meta: MacroMeta,
  handler: MacroHandler
): Macro {
  return {
    name,
    apply: handler,
    __types: renderMetaType(name, meta),
  } as InternalMacro
}

export function isMacro(o: unknown): o is InternalMacro {
  return !!(o as InternalMacro).__types
}

export function renderMetaType(name: string, meta: MacroMeta) {
  return [
    meta.types.join('\n'),
    meta.signatures
      .map((s) =>
        s.comment
          ? `  /** ${s.comment} **/
  export function ${name}${s.signature}`
          : `  export function ${name}${s.signature}`
      )
      .join('\n'),
  ]
    .filter((t) => !!t)
    .join('\n')
}
