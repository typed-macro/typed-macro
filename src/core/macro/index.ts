import { CallExpression, File, Program } from '@babel/types'
import { NodePath } from '@babel/traverse'

import { CURRENT_CALL_VERSION, Versioned } from '@/core/version'
import { State } from '../helper/state'
import { HELPER, MacroHelper } from './helper'
import { Babel, BABEL_TOOLS } from './babel'
import { ImportedMacro } from '@/core/helper/traverse'

export type MacroContext = {
  /**
   * The file path of the module currently being handled.
   */
  filepath: string
  /**
   * The original source code of the module currently being handled.
   */
  code: string
  /**
   * The arguments nodes of the call-macro expression currently being handled.
   */
  args: CallExpression['arguments']
  /**
   * The NodePath of the call-macro expression currently being handled.
   */
  path: NodePath<CallExpression>
  /**
   * The AST being traversed.
   */
  ast: File
  /**
   * Whether in ssr mode.
   */
  ssr: boolean
  /**
   * Traversal state.
   */
  state: {
    /**
     * In order to recursively expand all macros in a file,
     * the transformer will traverse the AST many times until all macros
     * are expanded.
     *
     * - {@link traversal} is the state shared during one traversal, and
     * - {@link transform} is the state shared during the whole transformation.
     */
    traversal: State
    /**
     * @see traversal
     */
    transform: State
  }
}

export type MacroHandler = (
  /**
   * Context about the macro caller.
   */
  ctx: MacroContext,
  /**
   * Babel tools to help operate node paths and nodes.
   */
  babel: Readonly<Babel>,
  /**
   * Wrappers upon babel tools to simplify operations.
   */
  helper: Readonly<MacroHelper>
) => void

type RawMacroCall = Versioned<{
  filepath: string
  code: string
  path: NodePath<CallExpression>
  ast: File
  program: NodePath<Program>
  ssr: boolean
  traversalState: State
  transformState: State
  importedMacros: ImportedMacro[]
}>

export type Macro = (call: RawMacroCall) => void

export type MacroMeta = {
  signatures: {
    comment?: string
    signature: string
  }[]
  types: string[]
}

type InternalMacro = Macro & {
  __types: string
}

export function macro(
  name: string,
  meta: MacroMeta,
  handler: MacroHandler
): Macro {
  const normalizer = getNormalizer(handler)
  const m = {
    [name](call: RawMacroCall) {
      const { ctx, babel, helper } = normalizer(call)
      handler(ctx, babel as Babel, helper as MacroHelper)
    },
  }[name]
  ;(m as InternalMacro).__types = renderMetaType(name, meta)
  return m
}

type MacroCallNormalizer = (raw: RawMacroCall) => {
  ctx: MacroContext
  babel?: Readonly<Babel>
  helper?: Readonly<MacroHelper>
}

function ensureCompatible(raw: RawMacroCall) {
  if (raw.__version !== CURRENT_CALL_VERSION)
    throw new Error('the macro is incompatible with the runtime')
}

function normalizeContext(raw: RawMacroCall): MacroContext {
  const { filepath, ssr, ast, code, path, transformState, traversalState } =
    raw.value
  return {
    filepath,
    code,
    ast,
    path,
    args: path.node.arguments,
    state: {
      transform: transformState,
      traversal: traversalState,
    },
    ssr,
  }
}

export function getNormalizer(handler: MacroHandler): MacroCallNormalizer {
  switch (handler.length) {
    case 0:
      throw new Error('not a valid macro handler')
    case 1:
      return (raw: RawMacroCall) => {
        ensureCompatible(raw)
        return {
          ctx: normalizeContext(raw),
        }
      }

    case 2:
      return (raw: RawMacroCall) => {
        ensureCompatible(raw)
        return {
          ctx: normalizeContext(raw),
          babel: BABEL_TOOLS,
        }
      }

    default:
      return (raw: RawMacroCall) => {
        ensureCompatible(raw)
        const ctx = normalizeContext(raw)
        const { path, filepath, program, importedMacros } = raw.value
        HELPER.props
          .setImportedMacros(importedMacros)
          .setPath(path)
          .setFilepath(filepath)
          .setProgram(program)
        return {
          ctx,
          babel: BABEL_TOOLS,
          helper: HELPER.instance,
        }
      }
  }
}

export function isMacro(o: unknown): o is InternalMacro {
  return (o as InternalMacro)?.__types !== undefined ?? false
}

export function renderMetaType(name: string, meta: MacroMeta) {
  return [
    meta.types.join('\n'),
    meta.signatures
      .map((s) =>
        s.comment
          ? `  /** ${s.comment} **/
  export function ${name}${s.signature}`
          : `  export function ${name}${s.signature}`
      )
      .join('\n'),
  ]
    .filter((t) => !!t)
    .join('\n')
}
