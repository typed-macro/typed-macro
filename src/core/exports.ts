import { isMacro, Macro } from '@/core/macro'
import { isMacroCompatible } from '@/core/compat'

export type NamespacedMacros = { [namespace: string]: Macro[] }

export type NamespacedModules = { [namespace: string]: string }

export type NamespacedTypes = {
  [namespace: string]: {
    moduleScope: string[]
    macroScope: string[]
  }
}

export type Exportable = (
  | {
      /**
       * as macros
       */
      macros: Macro[]
    }
  | {
      /**
       * as module, code will be used as source code of the module
       */
      code: string
    }
) & {
  /**
   * Type definitions, will be written to d.ts.
   *
   * e.g.
   * ```typescript
   * { exports: { '@macros': { customTypes: `type A = string` } } }
   * ```
   * will generate
   * ```typescript
   * declare module '@macros' {
   *   type A = string
   * }
   * ```
   */
  customTypes?: string
}

export type NamespacedExportable = { [namespace: string]: Exportable }

export type NormalizedExports = {
  macros: NamespacedMacros
  modules: NamespacedModules
  types: NamespacedTypes
}

export function normalizeExports(
  exports: NamespacedExportable
): NormalizedExports {
  const macros: NamespacedMacros = Object.create(null)
  const modules: NamespacedModules = Object.create(null)
  const types: NamespacedTypes = Object.create(null)
  Object.keys(exports).forEach((ns) => {
    const item = exports[ns]
    types[ns] = {
      moduleScope: item.customTypes ? [item.customTypes] : [],
      macroScope: [],
    }
    if ('code' in item) {
      modules[ns] = item.code
    } else {
      macros[ns] = item.macros
      types[ns].macroScope = item.macros.map((m) => {
        if (!isMacro(m)) throw new Error(`'${m.name}' is not a macro`)
        return m.__types
      })
    }
  })

  validateMacros(macros)

  return {
    macros,
    modules,
    types,
  }
}

// ensure compatible and no name conflict
export function validateMacros(macros: NamespacedMacros) {
  Object.keys(macros).forEach((ns) => {
    const mem = Object.create(null)
    macros[ns].forEach((m) => {
      if (!isMacroCompatible(m))
        throw new Error(`macro '${m.name}' in '${ns}' is incompatible`)

      if (mem[m.name])
        throw new Error(`macro '${m.name}' in '${ns}' is duplicated`)

      mem[m.name] = 1
    })
  })
}
