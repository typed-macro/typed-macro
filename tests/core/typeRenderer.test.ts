import {
  createTypeRenderer,
  renderTypes,
  TypeRenderer,
} from '@/core/typeRenderer'
import { NamespacedTypes } from '@/core/exports'
import { readFileSync, writeFileSync } from 'fs'
import { withTempPath } from '../testutils'

describe('renderTypes()', () => {
  it('should work', () => {
    expect(
      renderTypes({
        '@macro': {
          moduleScope: [`type A = string`],
          macroScope: [`export function c():A`],
        },
      })
    ).toMatchSnapshot()
    expect(
      renderTypes({
        '@macro': {
          moduleScope: [],
          macroScope: [`export function c():A`],
        },
      })
    ).toMatchSnapshot()
    expect(
      renderTypes({
        '@macro': {
          moduleScope: [],
          macroScope: [`export function c():A`],
        },
        '@another': {
          moduleScope: [`type B = string`],
          macroScope: [`export function d():B`],
        },
      })
    ).toMatchSnapshot()
  })
})

describe('TypeRenderer', () => {
  let types: NamespacedTypes
  let renderer: TypeRenderer
  beforeEach(() => {
    types = {}
    renderer = createTypeRenderer({
      types,
      typesPath: '',
    })
  })
  it('should render() properly', () => {
    expect(renderer.render()).toMatchSnapshot()
    types['@macro'] = {
      moduleScope: [`type A = string`],
      macroScope: [],
    }
    expect(renderer.render()).toMatchSnapshot()
  })

  it('should append() properly and update properly', () => {
    types['@macro'] = {
      moduleScope: [`type A = string`],
      macroScope: [],
    }
    {
      // append to existed namespace
      const update = renderer.append('@macro', 'type B = string')
      expect(renderer.render()).toMatchSnapshot()
      update(`type C = string`)
      expect(renderer.render()).toMatchSnapshot()
    }
    {
      // append to non-existed namespace
      const update = renderer.append('@another', 'type B = string')
      expect(renderer.render()).toMatchSnapshot()
      update(`type C = string`)
      expect(renderer.render()).toMatchSnapshot()
    }
  })
})

describe('TypeRenderer#write()', () => {
  const types = {
    '@macros': {
      moduleScope: [`type A = string`],
      macroScope: [],
    },
  }

  it('should ensure directory existed', (done) => {
    withTempPath('not/exist/a.d.ts', (tempPath) =>
      createTypeRenderer({
        typesPath: tempPath,
        types,
      })
        .write()
        .then(() => done())
    )
  })

  it('should truncate file', (done) => {
    withTempPath('./a.d.ts', (tempPath) => {
      writeFileSync(tempPath, 'hello\n'.repeat(10))
      const renderer = createTypeRenderer({
        typesPath: tempPath,
        types,
      })
      return renderer.write().then(() => {
        expect(readFileSync(tempPath).toString()).toBe(renderer.render())
        done()
      })
    })
  })
})
