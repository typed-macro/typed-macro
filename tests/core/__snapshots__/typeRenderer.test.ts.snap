// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeRenderer should append() properly and update properly 1`] = `
"declare module '@macro' {
type A = string
type B = string
}"
`;

exports[`TypeRenderer should append() properly and update properly 2`] = `
"declare module '@macro' {
type A = string
type C = string
}"
`;

exports[`TypeRenderer should append() properly and update properly 3`] = `
"declare module '@macro' {
type A = string
type C = string
}
declare module '@another' {
type B = string
}"
`;

exports[`TypeRenderer should append() properly and update properly 4`] = `
"declare module '@macro' {
type A = string
type C = string
}
declare module '@another' {
type C = string
}"
`;

exports[`TypeRenderer should render() properly 1`] = `""`;

exports[`TypeRenderer should render() properly 2`] = `
"declare module '@macro' {
type A = string
}"
`;

exports[`renderTypes() should work 1`] = `
"declare module '@macro' {
type A = string
export function c():A
}"
`;

exports[`renderTypes() should work 2`] = `
"declare module '@macro' {
export function c():A
}"
`;

exports[`renderTypes() should work 3`] = `
"declare module '@macro' {
export function c():A
}
declare module '@another' {
type B = string
export function d():B
}"
`;
